### DISCUSSION TITLE
LLDP packet drops on bond interface

### DISCUSSION THREAD

**Discussion by kstange-enfn** [lldpd/lldpd#644] [Q&A] [answered]
I've been having trouble finding much information on this, so I apologize if this is duplicated.  I have a bunch of Ubuntu 20.04 or 22.04 servers running kernel 5.15 or 6.2.  These servers have two intel NICs in bond, usually x710 hardware.

When `lldpd` is running, I see one dropped packet counted on the bond interface for each LLDP packet received by either physical interface.  They are processed correctly by `lldpd` and not shown as dropped on the physical interfaces.

When `lldpd` is _not_ running, drops are not counted on any interfaces.  If I run `tcpdump -i bond0 ether proto 0x88cc` I can see the packets reaching the bond.  I know they don't get counted when the interface is in promiscuous mode, but the drop counter just stops incrementing entirely as long as `lldpd` is stopped.

I tried using a tool called `dropwatch` and I see these only when `lldpd` is running:

```
drop at: __netif_receive_skb_core+0x150/0xf60 (0xffffffffacad1960)
origin: software
input port ifindex: 10
timestamp: Tue Mar 19 15:35:21 2024 706822244 nsec
protocol: 0x88cc
length: 322
original length: 322
```

These counters incrementing don't hurt anything, so I'm just trying to understand what's going on.  If there's a way to prevent it and keep these counters at zero, that's a bonus.

**Comment by vincentbernat** [comment]
See https://github.com/lldpd/lldpd/commit/8b50be7f61ad20ebae15372a509f7e778da2cc6f.

**Reply by kstange-enfn** [reply]
I'm doubtful this is the solution as I have lldpd 1.0.13 on Ubuntu 22.04 and lldpd 10.0.4 on Ubuntu 20.04 and see this behavior in both places.  I can try a newer lldpd if you think that will help.

**Reply by vincentbernat** [reply]
It's not a fix it's an explanation why this problem is there and not fixed.

**Reply by kstange-enfn** [reply]
I see, both the change and revert happened in between the lldpd versions we are using.  Is there no prospect for a workaround here, other than kernel behavior changes?  Just want to understand if there's any future for a fix.  If not then I'll just mark as resolved by this explanation and move on.

Thanks for your responses!

**Reply by vincentbernat** [reply, marked as answer]
No fix planned. I don't want to break some people setup "just" to fix an error counter. At the minimum, someone should test again if the patch breaks current versions of OVS.

**Comment by txj36** [comment]
I encountered the same problem on CentOS, with a kernel version of 6.6.47. I found that the content of /proc/net/ptype is different before and after starting the lldpd service with duplicate packet_rcv entries appearing.

On my device, there are two network interfaces eth2 and eth3, where eth3 is a slave interface of bond0.

1）before starting the lldpd service.
#cat /proc/net/ptype
Type Device      Function
...
ALL  eth2     packet_rcv
ALL  eth3     packet_rcv
0800          ip_rcv
0806          arp_rcv
86dd          ipv6_rcv
2）after starting the lldpd service.
#cat /proc/net/ptype
Type Device      Function
...
ALL  eth2     packet_rcv
ALL  eth2     packet_rcv
ALL  eth3     packet_rcv
ALL  eth3     packet_rcv
0800          ip_rcv
0806          arp_rcv
86dd          ipv6_rcv

In addition, the BPFtrace information output for _netif_receive_skb_core and kfree_skb is as follows:

```
181998.913560 netif_receive_skb_core pfmemalloc: 0, proto:0x88cc, skb addr: 0xffff00060e5f5600, rx_handler 0, packet type:2, skip_classify:0, skb->dev:eth2

        __netif_receive_skb_core+0
        __netif_receive_skb_list+288
        netif_receive_skb_list_internal+224
        napi_complete_done+108
        mlx5e_napi_poll+548
        __napi_poll+60
        napi_poll+340
        net_rx_action+216
        handle_softirqs+288
        __do_softirq+24
        ____do_softirq+20
        call_on_irq_stack+36
        do_softirq_own_stack+32
        irq_exit_rcu+248
        el1_interrupt+80
        el1h_64_irq_handler+32
        el1h_64_irq+120
        cpuidle_enter_state+192
        cpuidle_enter+60
        call_cpuidle+52
        cpuidle_idle_call+284
        do_idle+184
        cpu_startup_entry+56
        rest_init+204
        arch_call_rest_init+20
        start_kernel+924
        __primary_switched+188


181997.945443 netif_receive_skb_core pfmemalloc: 0, proto:0x88cc, skb addr: 0xffff00060e5f4000, rx_handler bond_handle_frame, packet type:2, skip_classify:0, skb->dev:eth3

        __netif_receive_skb_core+0
        __netif_receive_skb_list+288
        netif_receive_skb_list_internal+224
        napi_complete_done+108
        mlx5e_napi_poll+548
        __napi_poll+60
        napi_poll+340
        net_rx_action+216
        handle_softirqs+288
        __do_softirq+24
        ____do_softirq+20
        call_on_irq_stack+36
        do_softirq_own_stack+32
        irq_exit_rcu+248
        el1_interrupt+80
        el1h_64_irq_handler+32
        el1h_64_irq+120
        cpuidle_enter_state+192
        cpuidle_enter+60
        call_cpuidle+52
        cpuidle_idle_call+284
        do_idle+184
        cpu_startup_entry+60
        rest_init+204
        arch_call_rest_init+20
        start_kernel+924
        __primary_switched+188

181997.945476 kfree_skb>> skb addr 0xffff00060e5f4000, packet type: 2, drop_reaseon=51, location __netif_receive_skb_core protocol=0x88cc, skc_family 0 skb->dev:bond0
```
